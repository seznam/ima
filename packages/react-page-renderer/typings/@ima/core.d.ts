import { Utils } from '../../src/types';

declare module '@ima/core' {
  export class Cache {
    serialize(): void;
  }

  export class ComponentUtils {
    getUtils(): Utils;
  }

  export class Controller {
    beginStateTransaction(): void;
    commitStateTransaction(): void;
    getHttpStatus(): number;
    getState(): { [key: string]: unknown };
    setMetaParams(
      loadedResources: { [key: string]: unknown },
      metaManager: MetaManager,
      router: Router,
      dictionary: Dictionary,
      settings: { [key: string]: unknown }
    ): void;
    setState(statePatch: { [key: string]: unknown }): void;
    update(prevParams: { [key: string]: string }): {
      [key: string]: Promise<unknown> | unknown;
    };
  }

  export class ControllerDecorator extends Controller {
    getMetaManager(): MetaManager;
    setMetaParams(loadedResources: { [key: string]: unknown }): void;
  }

  export class Dictionary {
    get(
      key: string,
      parameters: { [key: string]: boolean | number | string | Date }
    ): string;
  }

  export class Dispatcher {
    fire(
      event: string,
      data: { [key: string]: unknown },
      imaInternalEvent: boolean = false
    ): void;
  }

  export class EventBus {
    fire(
      eventTarget: EventTarget,
      eventName: string,
      data: unknown,
      options?: { bubbles: boolean; cancelable: boolean }
    ): EventBus;
    listen(
      eventTarget: EventTarget,
      eventName: string,
      listener: (event: Event) => void
    ): EventBus;
    unlisten(
      eventTarget: EventTarget,
      eventName: string,
      listener: (event: Event) => void
    ): EventBus;
  }

  export class GenericError extends Error {
    constructor(
      message: string,
      params: { [key: string]: unknown } = {},
      dropInternalStackFrames: boolean = true
    );
    getHttpStatus(): number;
    getParams(): { [key: string]: unknown };
  }

  export class MetaManager {
    getLink(relation: string): { value: string; [key: string]: string };
    getLinks(): string[];
    getMetaName(name: string): { value: string; [key: string]: string };
    getMetaNames(): string[];
    getMetaProperties(): string[];
    getMetaProperty(name: string): { value: string; [key: string]: string };
    getTitle(): string;
    clearMetaAttributes(): void;
  }

  export class PageRenderer {
    mount(
      controller: Controller,
      view: ComponentType,
      pageResources: { [key: string]: unknown | Promise<unknown> },
      routeOptions: RouteOptions
    ): Promise<
      | unknown
      | {
          content?: string;
          pageState: { [key: string]: unknown };
          status: number;
        }
    >;
    update(
      controller: Controller,
      view: ComponentType,
      pageResources: { [key: string]: unknown | Promise<unknown> },
      routeOptions: RouteOptions
    ): Promise<
      | unknown
      | {
          content?: string;
          pageState: { [key: string]: unknown };
          status: number;
        }
    >;
    unmount(): void;
    setState(state: { [key: string]: unknown }): void;
  }

  export enum RendererEvents {
    MOUNTED = '$IMA.$PageRenderer.mounted',
    UPDATED = '$IMA.$PageRenderer.updated',
    UNMOUNTED = '$IMA.$PageRenderer.unmounted',
    ERROR = '$IMA.$PageRenderer.error',
  }

  export enum RendererTypes {
    RENDER = '$IMA.$PageRenderer.type.render',
    HYDRATE = '$IMA.$PageRenderer.type.hydrate',
    UNMOUNT = '$IMA.$PageRenderer.type.unmount',
  }

  export class Response {
    isResponseSent(): boolean;
    getResponseParams(): {
      status: number;
      content: string;
      pageState: { [key: string]: unknown };
    };
    send(content: string): Response;
    setPageState(pageState: { [key: string]: unknown }): Response;
    status(httpStatus: number): Response;
  }

  export class Router {
    link(name: string, params: { [key: string]: string | number }): string;
  }

  export class Window {
    getDocument(): Document;
    getElementById(id: string): Element;
    getHistoryState(): { [key: string]: unknown };
    querySelector(selector: string): Element;
    querySelectorAll(selector: string): Element[];
    getScrollX(): number;
    getScrollY(): number;
    getUrl(): string;
    getWindow(): undefined | globalThis.Window;
    pushState(
      state: { [key: string]: unknown },
      title: string | null,
      url: string
    ): void;
    replaceState(
      state: { [key: string]: unknown },
      title: string | null,
      url: string = null
    ): void;
    scrollTo(x: number, y: number): void;
    setTitle(title: string): void;
  }
}
