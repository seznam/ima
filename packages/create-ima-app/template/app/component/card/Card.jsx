import './card.less';

export default function Card({ children, title, href }) {
  return (
    <div className='card'>
      <a className='card-title' target='_blank' href={href} rel='noreferrer'>
        <h3>{title} &raquo;</h3>
      </a>
      <p
        className='card-content'
        dangerouslySetInnerHTML={{
          __html: children.replaceAll('{link}', href),
        }}
      />
    </div>
  );
}
