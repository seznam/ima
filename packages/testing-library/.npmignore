*
!dist/**/*
!jest-preset.js
!package.json
