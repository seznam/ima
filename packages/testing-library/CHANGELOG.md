# @ima/testing-library

## 19.11.1

### Patch Changes

- 5c50ea4: Fix Windows compatibility when using this package as a Jest preset

## 19.11.0

### Minor Changes

- 550793a9f: Add `renderHookWithContext`, see the [docs](https://imajs.io/basic-features/testing/#renderhookwithcontext) for more details.

## 19.10.0

### Minor Changes

- cd24e835d: Throw an error when JSDOM HTML template render failed. This can be potentially a **BREAKING CHANGE** if your tests are already using a broken HTML template in JSDOM. Until now, you might not have even noticed the problem until you had a test accessing specific context features that required a proper HTML template.

## 19.9.0

### Minor Changes

- 371ca1f20: Add client configuration options `beforeInitImaApp`, `beforeRenderWithContext` and `afterRenderWithContext` to be able to specify additional logic, which will be called in the specific times.

## 19.8.0

### Minor Changes

- 03d109ad6: Init new module, see README.md for more info.
