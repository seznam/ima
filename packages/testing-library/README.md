<p align="center">
  <img height="130" src="https://imajs.io/img/logo.svg">
</p>

<h1 align="center">@ima/testing-library</h1>
  <p align="center"><i>Testing library for IMA.js applications.</i>
</p>

---

## IMA Testing Library

The `@ima/testing-library` contains utilities for testing IMA.js applications. It provides integration with [Jest](https://jestjs.io), [React Testing Library](https://testing-library.com/docs/react-testing-library/intro) (RTL for short) and [Testing Library Jest DOM](https://testing-library.com/docs/ecosystem-jest-dom).

## Installation

Install the new dependencies. Note that RTL dependencies are only peer dependencies and you should specify them in your project.

```bash
npm install -D @ima/testing-library @testing-library/dom @testing-library/jest-dom @testing-library/react jest-environment-jsdom
```

Configure jest preset in your jest config file.

```json
{
  "preset": "@ima/testing-library"
}
```

Everything should start working out of the box for a typical IMA.js application. If you are trying to setup this library in a monorepo or an npm package, you might have to do some tweaks with the configuration.

### Configuration

There are 2 config functions that you can use to adjust the IMA Testing Library to your specific needs.

**Server Configuration**

In this case, you need the jest config file to be in non-json format.

This configuration should be evaluated in the jest config file. It's config values are used to initialize the JSDOM environment in which the tests are running.

```javascript
const path = require('node:path');
const { setImaTestingLibraryServerConfig } = require('@ima/testing-library/server');

setImaTestingLibraryServerConfig({
  // your custom config
  applicationFolder: path.resolve('./__tests__/') // The default application folder is the root of the project, but you can specify a custom one to add some test specific logic.
});

module.exports = {
  preset: '@ima/testing-library',
  // The preset automatically sets up the moduleNameMapper for the IMA.js application, but you can override it if you need to.
  moduleNameMapper: {
    '^app/main$': '<rootDir>/app/main.test.js', // You can tell jest to use a different main file for the tests
  }
};
```

See [src/server/configuration.ts](https://github.com/seznam/ima/blob/master/packages/testing-library/src/server/configuration.ts) for the full list of available options.

**Client Configuration**

This configuration should be evaluated in the setup files, or directly in the test files. It's config values are used to initialize the IMA.js application and provide the context for the tests.

```javascript
const { setImaTestingLibraryClientConfig } = require('@ima/testing-library/client');

setImaTestingLibraryClientConfig({
  // your custom config
  imaConfigPath: 'path/to/your/ima.config.js',
});
```

See [src/client/configuration.ts](https://github.com/seznam/ima/blob/master/packages/testing-library/src/client/configuration.ts) for the full list of available options.

## Usage

See [documentation](https://imajs.io/basic-features/testing) for more information about how to use the IMA Testing Library.
