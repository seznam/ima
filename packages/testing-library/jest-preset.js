// This file is neccessary for Windows users
// https://github.com/jestjs/jest/issues/15511
module.exports = require('./dist/cjs/jest-preset.js');
