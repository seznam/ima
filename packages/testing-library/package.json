{
  "name": "@ima/testing-library",
  "version": "19.11.0",
  "description": "Testing library for IMA.js applications.",
  "keywords": [
    "IMA.js",
    "react",
    "testing",
    "library"
  ],
  "bugs": {
    "url": "https://github.com/seznam/ima/issues"
  },
  "repository": {
    "type": "git",
    "url": "https://github.com/seznam/ima.git",
    "directory": "packages/testing-library"
  },
  "license": "MIT",
  "author": "Filip Satek <filip.satek@firma.seznam.cz>",
  "scripts": {
    "dev": "node ../plugin-cli/dist/bin/ima-plugin.js dev",
    "build": "node ../plugin-cli/dist/bin/ima-plugin.js build",
    "link": "node ../plugin-cli/dist/bin/ima-plugin.js link",
    "lint": "eslint './**/*.{js,jsx,ts,tsx}'",
    "test": "jest -c jest.config.js"
  },
  "exports": {
    ".": {
      "types": "./dist/esm/index.d.ts",
      "import": "./dist/esm/index.js",
      "default": "./dist/cjs/index.js"
    },
    "./client": {
      "types": "./dist/esm/client/index.d.ts",
      "import": "./dist/esm/client/index.js",
      "default": "./dist/cjs/client/index.js"
    },
    "./server": {
      "types": "./dist/esm/server/index.d.ts",
      "import": "./dist/esm/server/index.js",
      "default": "./dist/cjs/server/index.js"
    },
    "./fallback/app/main": {
      "types": "./dist/esm/client/app/main.d.ts",
      "import": "./dist/esm/client/app/main.js",
      "default": "./dist/cjs/client/app/main.js"
    },
    "./fallback/server/*": {
      "types": "./dist/esm/server/*.d.ts",
      "import": "./dist/esm/server/*.js",
      "default": "./dist/cjs/server/*.js"
    },
    "./jest-preset": {
      "types": "./dist/esm/jest-preset.d.ts",
      "import": "./dist/esm/jest-preset.js",
      "default": "./dist/cjs/jest-preset.js"
    },
    "./jestSetupFileAfterEnv": {
      "types": "./dist/esm/jestSetupFileAfterEnv.d.ts",
      "import": "./dist/esm/jestSetupFileAfterEnv.js",
      "default": "./dist/cjs/jestSetupFileAfterEnv.js"
    }
  },
  "dependencies": {
    "@messageformat/core": "^3.0.1",
    "globby": "^11.1.0"
  },
  "peerDependencies": {
    "@ima/cli": ">=19.0.0",
    "@ima/core": ">=19.0.0",
    "@ima/server": ">=19.0.0",
    "@ima/react-page-renderer": ">=19.0.0",
    "@testing-library/dom": ">=10.0.0",
    "@testing-library/jest-dom": ">=6.0.0",
    "@testing-library/react": ">=12.0.0",
    "express": ">=4.0.0"
  },
  "publishConfig": {
    "access": "public",
    "registry": "https://registry.npmjs.org/"
  }
}
