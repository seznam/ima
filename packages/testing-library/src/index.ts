export * from './rtl';
export * from './types';
