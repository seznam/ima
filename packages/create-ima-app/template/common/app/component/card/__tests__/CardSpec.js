import { renderWithContext } from '@ima/testing-library';

import { Card } from '../Card';

describe('Card', () => {
  it('can render', async () => {
    const props = {
      title: 'Test card',
      href: 'https://www.seznam.cz',
      children: 'Some content of the card.',
    };

    const { container } = await renderWithContext(<Card {...props} />);

    expect(container.firstChild).toMatchSnapshot();
  });
});
