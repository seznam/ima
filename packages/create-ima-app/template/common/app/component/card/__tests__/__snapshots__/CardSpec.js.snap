// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Card can render 1`] = `
<div
  class="card"
>
  <a
    class="card-title"
    href="https://www.seznam.cz"
    rel="noreferrer"
    target="_blank"
  >
    <h3>
      Test card
       Â»
    </h3>
  </a>
  <p
    class="card-content"
  >
    Some content of the card.
  </p>
</div>
`;
